File: j-LIB-core/src/main/java/io/github/dot166/jLib/app/jLibViewInflater.java
package io.github.dot166.jLib.app;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SwitchCompat;
import androidx.appcompat.widget.Toolbar;

import com.google.android.material.switchmaterial.SwitchMaterial;
import com.google.android.material.theme.MaterialComponentsViewInflater;

import org.jetbrains.annotations.Contract;

import io.github.dot166.jLib.widget.ActionBar2;

public class jLibViewInflater extends MaterialComponentsViewInflater {

    @Override
    @Nullable
    protected View createView(Context context, @NonNull String name, AttributeSet attrs) {
        View view = null;

        // We need to 'inject' our Views in place of the standard framework versions and appcompat versions and materialcomponents versions
        switch (name) {
            case "com.google.android.material.materialswitch.MaterialSwitch": // does not theme properly so replace with the older one
                view = createSwitch(context, attrs);
                verifyNotNull(view, name);
                break;
            case "androidx.appcompat.widget.SwitchCompat":
                view = createSwitch(context, attrs);
                verifyNotNull(view, name);
                break;
            case "Switch":
                view = createSwitch(context, attrs);
                verifyNotNull(view, name);
                break;
            case "Toolbar":
                view = createToolbar(context, attrs);
                verifyNotNull(view, name);
                break;
            case "androidx.appcompat.widget.Toolbar":
                view = createToolbar(context, attrs);
                verifyNotNull(view, name);
                break;
            case "com.google.android.material.appbar.MaterialToolbar":
                view = createToolbar(context, attrs);
                verifyNotNull(view, name);
                break;
            default:
                return super.createView(context, name, attrs);
        }
        return view;
    }

    @NonNull
    @Contract("_, _ -> new")
    private Toolbar createToolbar(Context context, AttributeSet attrs) {
        return new ActionBar2(context, attrs);
    }

    @NonNull
    @Contract("_, _ -> new")
    protected SwitchCompat createSwitch(Context context, AttributeSet attrs) {
        return new SwitchMaterial(context, attrs);
    }

    private void verifyNotNull(View view, String name) {
        if (view == null) {
            throw new IllegalStateException(this.getClass().getName()
                    + " asked to inflate view for <" + name + ">, but returned null");
        }
    }
}

File: j-LIB-core/src/main/java/io/github/dot166/jlib/app/jLibViewInflater.kt
package io.github.dot166.jlib.app

import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.view.View
import androidx.appcompat.widget.SwitchCompat
import androidx.appcompat.widget.Toolbar
import androidx.cardview.widget.CardView
import com.google.android.material.appbar.MaterialToolbar
import com.google.android.material.card.MaterialCardView
import com.google.android.material.materialswitch.MaterialSwitch
import com.google.android.material.theme.MaterialComponentsViewInflater
import org.jetbrains.annotations.Contract

open class jLibViewInflater : MaterialComponentsViewInflater() {
    override fun createView(context: Context, name: String, attrs: AttributeSet?): View? {
        var view: View? = null

        // We need to 'inject' our Views in place of the standard framework versions and appcompat versions and materialcomponents versions
        when (name) {
            "com.google.android.material.switchmaterial.SwitchMaterial" -> {
                view = createSwitch(context, attrs)
                verifyNotNull(view, name)
            }

            "androidx.appcompat.widget.SwitchCompat" -> {
                view = createSwitch(context, attrs)
                verifyNotNull(view, name)
            }

            "androidx.appcompat.widget.Toolbar" -> {
                view = createToolbar(context, attrs)
                verifyNotNull(view, name)
            }

            "androidx.cardview.widget.CardView" -> {
                view = createCardView(context, attrs)
                verifyNotNull(view, name)
            }

            else -> return super.createView(context, name, attrs)
        }
        Log.i(this.javaClass.simpleName, view.javaClass.name)
        return view
    }

    @Contract("_, _ -> new")
    protected fun createToolbar(context: Context, attrs: AttributeSet?): Toolbar {
        return MaterialToolbar(context, attrs)
    }

    @Contract("_, _ -> new")
    protected fun createSwitch(context: Context, attrs: AttributeSet?): SwitchCompat {
        return MaterialSwitch(context, attrs)
    }

    @Contract("_, _ -> new")
    protected fun createCardView(context: Context?, attrs: AttributeSet?): CardView {
        return MaterialCardView(context, attrs)
    }

    private fun verifyNotNull(view: View, name: String?) {
        checkNotNull(view) {
            (this.javaClass.getName()
                    + " asked to inflate view for <" + name + ">, but returned null")
        }
    }
}
